source/common/http/path_utility.cc

```diff
@@ -52,5 +52,43 @@ bool PathUtil::canonicalPath(HeaderEntry& path_header) {
   return true;
 }
 
+/* static */
+void PathUtil::mergeSlashes(HeaderEntry& path_header) {
+  const auto original_path = path_header.value().getStringView();
+  if (original_path.empty()) {
+    return;
+  }
+
+  bool has_adjacent_slashes = false;
+  for (size_t i = 1; i < original_path.size(); ++i) {
+    if (original_path[i] == '/' && original_path[i - 1] == '/') {
+      has_adjacent_slashes = true;
+      break;
+    }
+    // Only operate on path component in URL.
+    if (original_path[i] == '?') {
```

Comment:
I think we may also need to handle URLs ending with `#` components but without any query parameters.

Reply:
`mutateResponseHeaders` above explicitly treats only query part as special, is it a bug or intended?

Reply:
I'd retain characters after a `#` as-is i.e. not merge slashes.

Reply:
I agree with the statement, but this comment makes me believe that `canonicalPath` can be wrong as well:
https://github.com/envoyproxy/envoy/blob/d44c93a3fcb2e970829051b1a2f39800f652c370/source/common/http/path_utility.cc#L32-L38

Reply:
Okay, then please ignore my url fragment (`#`) related comments. We can revisit this at a later point if needed.

Reply:
Sounds good, I've addressed all other ones.